include/miroil/miroil/mirbuffer.h

```diff
@@ -21,41 +21,38 @@
 #include <mir/version.h>
 #include <memory>
 
-#include <GL/gl.h>
-
-namespace mir { namespace graphics { class Buffer; }}
+namespace mir {
+    namespace graphics {
+        class Buffer;
+
+        namespace gl {
+            class Texture;
+        }
+    }
+}
 
 namespace miroil
 {
 class GLBuffer
 {
 public:
-    GLBuffer();
     ~GLBuffer();
     explicit GLBuffer(std::shared_ptr<mir::graphics::Buffer> const& buffer);
 
-    operator bool() const;
     bool has_alpha_channel() const;
     mir::geometry::Size size() const;
+    void bind();
 
-    void reset();
     void reset(std::shared_ptr<mir::graphics::Buffer> const& buffer);
-    void bind();
-#if MIR_SERVER_VERSION < MIR_VERSION_NUMBER(2, 3, 0)
-    void gl_bind_tex();
-#endif
 
-private:
-    void init();
-    void destroy();
+    void reset();
+    bool empty();
+
+    static std::shared_ptr<GLBuffer> from_mir_buffer(std::shared_ptr<mir::graphics::Buffer> const& buffer);
```

Comment:
Why return a `shared_ptr<>`? This looks like a class with valie semantics.

Reply:
what do you mean? We do return shared_ptr?

Reply:
We can return a normal without, but we use shared_ptr upstream wherever we can. I can change this if needed?

Reply:
It is needless inefficiency. Passing `GLBuffer` around by value is more in keeping with its interface.

Reply:
In upstream qtmir its needed as we have to pass around as a virtual parent. so we need to pass it was a pointer.

Reply:
1. The class is clearly not designed as a base class, so why?
2. This function isn't constructing a QtMir type, so that doesn't follow.

Reply:
Here in mir its not as i removed that, but in qtmir it its https://gitlab.com/ubports/development/core/qtmir/-/blob/f04dbcff29e7cb7e7cd84749e30028d00ba1ddbb/src/platforms/mirserver/miral/mirbuffer.h

Reply:
It will need some ifdefs in qtmir to support both mir1 and mir2, but that part is not done yet

Reply:
also https://gitlab.com/ubports/development/core/qtmir/-/blob/f04dbcff29e7cb7e7cd84749e30028d00ba1ddbb/src/platforms/mirserver/miral/mirbuffer.cpp#L35

Reply:
_If_ compatibility code really needs a `shared_ptr<GLBuffer>`, then it is easy to write:
```c++
#if ...
   auto bufptr = std::make_shared<GLBuffer>(GLBuffer::from_mir_buffer(buf));
#else
   auto bufptr = GLBuffer::from_mir_buffer(buf);
#endif
```
But, as I recall (not looked at the code for ages), I had QtMir working without this.

https://github.com/MirOil-for-Lomiri/qtmir

Reply:
Would it be ok to change this later on when we drop mir1 support? I want to keep the delta for now with least amount if #ifdefs to keep it simpler

Reply:
How about we deprecate it now and drop it later?

There's nothing it does that isn't better done by the constructor.

Reply:
Yeah make sense, will do that.

Reply:
Added  now :)